File: Cpp/Intermediate/boj_1167.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

/*
[ 트리 지름 구하기 ]
 임의의 정점 v1, v1에서 가장 먼 정점 v2, v2에서 가장 먼 정점 v3라 할 때,
 v2와 v3를 연결하면 트리의 지름이 된다.
 */

constexpr int MAX = 100001;
static int max_distance = 0;
static int target = 0;
static bool visited[MAX];
static vector<pair<int, int>> tree[MAX];

/* node는 임의의 정점, sum은 거리 합 */
static void dfs(int node, int sum) {
    if (visited[node])
        return;

    if (max_distance < sum) {
        max_distance = sum;
        target = node;
    }

    visited[node] = true;

    for (int i = 0; i < tree[node].size(); i++) {
        int next = tree[node][i].first;
        int weight = tree[node][i].second;

        if (!visited[next])
            dfs(next, sum + weight);
    }
}

static void solve() {
    /* 이 함수가 종료되면 target에는 임의의 정점 1(v1)로부터 가장 먼 노드(v2)가 설정, max_distance에는 v1과 v2 사이 최대 거리 값이 설정됨*/
    dfs(2, 0);

    fill(visited, visited + MAX, false);

    /* 이 함수가 종료되면 target에는 v2에서 가장 먼 노드가, max_distance에는 가장 긴 거리값이 설정됨 */
    dfs(target, 0);

    cout << max_distance;
}

void boj_1167() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int v;
    cin >> v;

    while (v--) {
        int u;
        cin >> u;

        while (1) {
            int v;
            cin >> v;

            if (v == -1) break;

            int w;
            cin >> w;

            tree[u].push_back(make_pair(v, w));
        }
    }

    solve();
}

File: Kotlin/src/main/kotlin/intermediate/boj_1167.kt
package intermediate

private const val MAX = 100001
private val grp = Array<MutableList<Pair<Int, Int>>>(MAX) { mutableListOf() }
private val visited = BooleanArray(MAX) { false }
private var target = -1
private var max_dist = -1

private fun dfs(node: Int, sum: Int) {
    visited[node] = true

    if (max_dist < sum) {
        max_dist = sum
        target = node
    }

    for ((next, weight) in grp[node])
        if (!visited[next]) dfs(next, sum + weight)
}

private fun solve(): Int {
    dfs(1, 0)
    visited.fill(false)
    dfs(target, 0)
    return max_dist
}

fun boj_1167() = with(System.`in`.bufferedReader()) {
    val v = readLine().toInt()

    repeat (v) {
        val tmp = readLine()
            .split(" ")
            .map { it.toInt() }
            .toIntArray()

        val node = tmp[0]
        var idx = 1

        while (true) {
            if (tmp[idx] == -1) break
            grp[node].add(Pair(tmp[idx], tmp[idx + 1]))
            idx += 2
        }
    }

    print(solve())
}


/*

data class pr(val index: Int, val weight: Int): Comparable<pr> {
    override fun compareTo(other: pr): Int {
        return this.weight - other.weight
    }
}

fun dijk(grp: Array<MutableList<Pair<Int, Int>>>, start: Int): Pair<Int, Int> {
    val pq = PriorityQueue<pr>()
    val distance = IntArray(grp.size){ Int.MAX_VALUE }

    pq.add(pr(0, start))
    distance[start] = 0

    while (pq.isNotEmpty()) {
        val (dist, now) = pq.poll()

        if (distance[now] < dist)
            continue

        for ((next, weight) in grp[now]) {
            val cost = distance[now] + weight

            if (distance[next] > cost) {
                distance[next] = cost
                pq.add(pr(cost, next))
            }
        }
    }

    var rst = -1
    var rst_index = -1

    for ((i, k) in distance.withIndex()) {
        if (k == Int.MAX_VALUE) continue

        if (k > rst) {
            rst = k
            rst_index = i
        }
    }

    return Pair(rst, rst_index)
}
 */